WalletWasabi/Services/SmartCoinSelector.cs

```diff
@@ -0,0 +1,89 @@
+using NBitcoin;
+using System.Linq;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using WalletWasabi.Models;
+using WalletWasabi.Exceptions;
+
+namespace WalletWasabi.Services
+{
+	public class SmartCoinSelector : ICoinSelector
+	{
+		private Dictionary<OutPoint, SmartCoin> _smartsCoinByOutpoint;
+
+		public SmartCoinSelector(Dictionary<OutPoint, SmartCoin> smartCoinsByOutpoint)
+		{
+			_smartsCoinByOutpoint = smartCoinsByOutpoint;
+		}
+
+		public IEnumerable<ICoin> Select(IEnumerable<ICoin> coins, IMoney target)
+		{
+			var coinsByOutpoint = coins.ToDictionary(c => c.Outpoint);
+			var totalOutAmount = (Money)target;
+			var unspentCoins = coins.Select(c => _smartsCoinByOutpoint[c.Outpoint]).ToArray();
+			var coinsToSpend = new HashSet<SmartCoin>();
+			var unspentConfirmedCoins = new List<SmartCoin>();
+			var unspentUnconfirmedCoins = new List<SmartCoin>();
+			foreach (SmartCoin coin in unspentCoins)
+			{
+				if (coin.Confirmed)
+				{
+					unspentConfirmedCoins.Add(coin);
+				}
+				else
+				{
+					unspentUnconfirmedCoins.Add(coin);
+				}
+			}
+
+			bool haveEnough = TrySelectCoins(ref coinsToSpend, totalOutAmount, unspentConfirmedCoins);
+			if (!haveEnough)
+			{
+				haveEnough = TrySelectCoins(ref coinsToSpend, totalOutAmount, unspentUnconfirmedCoins);
+			}
+
+			if (!haveEnough)
+			{
+				throw new InsufficientBalanceException(totalOutAmount, unspentConfirmedCoins.Select(x => x.Amount).Sum() + unspentUnconfirmedCoins.Select(x => x.Amount).Sum());
+			}
+
+			return coinsToSpend.Select(c => coinsByOutpoint[c.GetOutPoint()]);
+		}
+
+		/// <returns>If the selection was successful. If there's enough coins to spend from.</returns>
+		private bool TrySelectCoins(ref HashSet<SmartCoin> coinsToSpend, Money totalOutAmount, IEnumerable<SmartCoin> unspentCoins)
+		{
+			// If there's no need for input merging, then use the largest selected.
+			// Do not prefer anonymity set. You can assume the user prefers anonymity set manually through the GUI.
+			SmartCoin largestCoin = unspentCoins.OrderByDescending(x => x.Amount).FirstOrDefault();
+			if (largestCoin == default)
+			{
+				return false; // If there's no coin then unsuccessful selection.
+			}
+			else // Check if we can do without input merging.
+			{
+				if (largestCoin.Amount >= totalOutAmount)
+				{
+					coinsToSpend.Add(largestCoin);
+					return true;
+				}
+			}
+
+			// If there's a need for input merging.
+			foreach (var coin in unspentCoins
+				.OrderByDescending(x => x.AnonymitySet) // Always try to spend/merge the largest anonset coins first.
+				.ThenByDescending(x => x.Amount)) // Then always try to spend by amount.
+			{
+				coinsToSpend.Add(coin);
+				// If reaches the amount, then return true, else just go with the largest coin.
+				if (coinsToSpend.Select(x => x.Amount).Sum() >= totalOutAmount)
```

Comment:
I think it could be more readable in this way (but it is not really important, btw):

```suggestion
				if (coinsToSpend.Sum(x => x.Amount) >= totalOutAmount)
```

Reply:
If well this will never be a problem for performance, it worth to note that summing inside the `foreach` loop can be eliminated by using an accumulator.

Reply:
See https://github.com/zkSNACKs/WalletWasabi/pull/2116/files/eee13f98b226d65e0f89eabd5778d919a768481f#r314208591